feat: Add MaxMovedRedirections option to prevent infinite redirect loops

nirchetrit opened the pull request:
## Description
This PR introduces a new `ClusterOption` called `MaxMovedRedirections`. This option allows users to configure a maximum limit for `MOVED` | `ASK` redirections during cluster operations.

Previously, the client could potentially enter an infinite loop of redirects if the cluster was misconfigured or unstable. This change ensures the client fails fast with a specific error (`ErrMaxMovedRedirectionsExceeded`) once the configured redirection limit is reached.

## Changes
- **New Option**: Added `MaxMovedRedirections` to `ClusterOption`.
- **Logic Update**: Updated the retry loops in `cluster.go` (including `DoMulti` and pipelines) to respect this limit.
- **Error Handling**: Enhanced error inspection to ensure we only count actual `MOVED` | `ASK` responses against this limit, preserving other errors (like network issues) where appropriate.

## Motivation
To allow the client to fail faster with a specific error in case of a moved redirection loop during cluster rebalance/topology changes.

nirchetrit committed:
implement MovedRetries

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
- Changed module path from github.com/valkey-io/valkey-go to github.com/nirchetrit/valkey-go
- Added dependency for github.com/valkey-io/valkey-go v1.0.67
- Removed indirect dependencies for github.com/kr/pretty and github.com/kr/text

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
update module path from github.com/nirchetrit/valkey-go to github.com/valkey-io/valkey-go

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
added MaxMovedRetries to the readme

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
Implement MaxMovedRetries logic in DoMulti and DoMultiCache methods, ensuring proper error handling and retry behavior. Update README to clarify the behavior of MaxMovedRetries for multi-command operations.

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
Update module path from github.com/valkey-io/valkey-go to github.com/nirchetrit/valkey-go

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
Refactor MaxMovedRetries handling in DoMulti and DoMultiCache methods to use a local movedRetries counter. This improves error handling for exceeded retry limits and ensures consistent behavior across retry attempts.

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
Update module path from github.com/nirchetrit/valkey-go to github.com/valkey-io/valkey-go and remove outdated dependencies. Adjust golang.org/x/tools version to v0.28.0.

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
Revert changes to README.md

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
Rename MaxMovedRetries to MaxMovedRedirections and update related error handling in cluster and test files for consistency.

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
Enhance error handling for MaxMovedRedirections in cluster.go by refining the condition checks for ValkeyError. This ensures that errors related to moved keys are correctly identified and handled during retries.

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
Clarify comments in cluster.go regarding error handling for MOVED responses during retries. This enhances code readability and maintains consistency with previous error handling updates.

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

rueian commented on cluster.go:
```diff
@@ -1259,7 +1297,24 @@ retry:
+						if err := results.s[i].val.Error(); err != nil {
+							if vErr, ok := err.(*ValkeyError); ok {
+								if _, isMoved := vErr.IsMoved(); isMoved {
+									results.s[i] = newErrResult(ErrMaxMovedRedirectionsExceeded)
```
Overall looks good to me, but why don't we just keep the original errors? I am not sure if hiding the redirect information is a good idea.

And should we also apply this configuration to ASK redirection as well?

nirchetrit replied:
I changed it to return the last received redirect error (we now include both ask and moved in the counter)

nirchetrit committed:
Rename MaxMovedRedirections to MaxRedirects and handle ASK redirections. expose the last received error instead of a generic max moved errors exceeded

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

nirchetrit committed:
Add tests for handling ASK and MOVED errors in cluster client

This commit introduces multiple test cases in cluster_test.go to validate the behavior of the cluster client when exceeding the maximum allowed ASK retries and handling mixed MOVED and ASK errors. The tests ensure that the client correctly identifies and returns ValkeyError for these scenarios, enhancing the robustness of error handling in the client.

The new tests cover:
- Exceeding max ASK retries for Do, DoCache, DoMulti, and DoMultiCache methods.
- Handling mixed MOVED and ASK errors during retries.

These additions improve the overall reliability of the cluster client in error-prone situations.

Signed-off-by: nirchetrit <nir.chetrit@gmail.com>

rueian committed:
revert go.mod

Signed-off-by: Rueian <rueiancsie@gmail.com>

rueian commented:
Thank you @nirchetrit!
